File: dotnet/src/Hl7v2ToFhirEhdsAdapter.WebApi/Controllers/ConvertController.cs
using Hl7.Fhir.Serialization;
using Hl7v2ToFhirEhdsAdapter.Conversion;
using Microsoft.AspNetCore.Mvc;

namespace Hl7v2ToFhirEhdsAdapter.WebApi.Controllers;

/// <summary>
/// API controller for HL7v2 to FHIR R4 conversion.
/// </summary>
[ApiController]
[Route("api/v1/convert")]
public class ConvertController : ControllerBase
{
    private readonly IConversionService _conversionService;
    private readonly FhirJsonSerializer _serializer;

    /// <summary>
    /// Initializes a new instance of the <see cref="ConvertController"/> class.
    /// </summary>
    /// <param name="conversionService">The conversion service.</param>
    public ConvertController(IConversionService conversionService)
    {
        _conversionService = conversionService ?? throw new ArgumentNullException(nameof(conversionService));
        _serializer = new FhirJsonSerializer(new SerializerSettings { Pretty = true });
    }

    /// <summary>
    /// Converts a raw HL7v2 message to a FHIR R4 JSON bundle.
    /// </summary>
    /// <param name="hl7Message">The raw HL7v2 message.</param>
    /// <returns>The JSON-serialized FHIR bundle.</returns>
    [HttpPost]
    [Consumes("application/x-www-form-urlencoded", "text/plain")]
    [Produces("application/json")]
    public async Task<IActionResult> Convert([FromForm] string hl7Message)
    {
        var bundle = _conversionService.Convert(hl7Message);
        string json = await _serializer.SerializeToStringAsync(bundle).ConfigureAwait(false);
        return Content(json, "application/json");
    }
}

File: src/main/java/ie/hl7/ehds/adapter/web/ConvertController.java
package ie.hl7.ehds.adapter.web;

import ca.uhn.fhir.context.FhirContext;
import ie.hl7.ehds.adapter.convert.ConversionService;
import org.hl7.fhir.r4.model.Bundle;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * REST controller for HL7v2 to FHIR R4 conversion.
 */
@RestController
@RequestMapping("/api/v1/convert")
public class ConvertController {

    private final ConversionService conversionService;
    private final FhirContext fhirContext;

    /**
     * Creates the controller with required dependencies.
     *
     * @param conversionService the conversion service
     */
    public ConvertController(final ConversionService conversionService) {
        this.conversionService = conversionService;
        this.fhirContext = FhirContext.forR4();
    }

    /**
     * Converts a raw HL7v2 message to a FHIR R4 JSON bundle.
     *
     * @param hl7Message the raw HL7v2 message
     * @return the JSON-serialized FHIR bundle
     */
    @PostMapping(produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<String> convert(@RequestBody final String hl7Message) {
        final Bundle bundle = conversionService.convert(hl7Message);
        final String json = fhirContext.newJsonParser().setPrettyPrint(true).encodeResourceToString(bundle);
        return ResponseEntity.ok(json);
    }
}
